spec/lib/rex/proto/LDAP/server_spec.rb

```diff
@@ -0,0 +1,147 @@
+        expect(server.running?).to be true
+      end
+
+      after { server.stop }
+    end
+  end
+
+  describe '#start' do
+    context 'start server with the provided options' do
+      before { server.start }
+
+
+      it 'starts the UDP server if serve_udp is true' do
+        if server.serve_udp
+          expect(server.udp_sock).to be_a(Rex::Socket::Udp)
+          expect(server.running?).to be true
+        end
+      end
+
+      it 'starts the TCP server if serve_tcp is true' do
+        if server.serve_tcp
+          expect(server.tcp_sock).to be_a(Rex::Socket::TcpServer)
+          expect(server.running?).to be true
+        end
+      end
+
+      after { server.stop }
+    end
+  end
+
+  describe '#stop' do
+    before { server.start }
+
+    it 'stops the server when running' do
+      server.stop
+      expect(server.running?).to be false
+    end
+  end
+
+  describe '#dispatch_request' do
+    it 'calls dispatch_request_proc if it is set' do
+      client = double('client')
+      block_called = false
+      server.dispatch_request_proc = proc { block_called = true }
+      server.dispatch_request(client, 'LDAP request data')
+      expect(block_called).to be true
+    end
+
+    it 'calls default_dispatch_request if dispatch_request_proc is not set' do
+      client = double('client')
+      expect { server.dispatch_request(client, 'LDAP request data') }.not_to raise_error
+    end
+  end
+
+  describe '#default_dispatch_request' do
+    it 'returns nil for empty request data' do
+      client = double('client')
+      data = ''
+      expect { server.default_dispatch_request(client, data) }.not_to raise_error
+    end
+
+    it 'closes the client for unsupported request types' do
+      client = double('client')
+      data = 'InvalidRequestData'
+      expect(client).to receive(:close)
+      server.default_dispatch_request(client, data)
+    end
+  end
+
+  describe '#encode_ldap_response' do
+    it 'encodes an LDAP response correctly' do
+      msgid = 1
+      code = Net::LDAP::ResultCodeSuccess
+      dn = ''
+      msg = Net::LDAP::ResultStrings[Net::LDAP::ResultCodeSuccess]
+      tag = Net::LDAP::PDU::BindResult
+      context_data = nil
+      context_code = nil
+
+      response = server.encode_ldap_response(msgid, code, dn, msg, tag, context_data, context_code)
+      expect(response).to be_a(String)
+    end
+  end
+
+  describe '#search_result' do
+    context 'when searching with no LDIF data' do
+      it 'returns a random search result' do
+        result = server.search_result(nil, 1)
+
+        expect(result).to be_a(String)
+      end
+    end
+  end
+
+  describe '#processed_pdu_handler' do
+    it 'sets the processed_pdu_handler correctly' do
+      server.processed_pdu_handler(Net::LDAP::PDU::BindRequest) do |processed_data|
+        processed_data = 'Processed Data'
+      end
+
+      expect(server.instance_variable_get(:@pdu_process)[Net::LDAP::PDU::BindRequest]).to be_a(Proc)
```

Comment:
Consider an additional test that actually proves the proc will be called if a BindRequest is received.

This also raises the question of what is expected to happen if the `proc` raises an exception?

Reply:
@jmartin-tech Any idea how could the calling of the `proc` be tested as it is performed in the `default_dispatch_request` and not as a seperate method

Reply:
also regarding exception I assume that should be handled by the module when describing what to do with the processed data.

Reply:
To test the proc you would need to exercise the call, this could be done in a few pattern, the simplest may be

``` Ruby
    found_proc = server.instance_variable_get(:@pdu_process)[Net::LDAP::PDU::BindRequest])
    expect(found_proc.call({})).to eq('Processed Data')
```

This would validate the `Proc` preformed the registered task. Although this would not validate that the `default_dispatch_request` is passing the correct data to the proc.

Another way would be to mock the client request on the socket with a simple auth `BindRequest` and intercept the call to `process_login_request` returning a known hash and registering a proc that will simply return the parameter that is passed into it for comparison in the test.

As to the exception handling, that could be noted in the Yard docs, as that would put the onus on the module writer. I considered the idea of  adding a `rescue` to the if statement for post processing and logging the error occurred, however the more I think about it the current behavior that would skip responding and close the client connection, log, and raise the error higher for the user to to see is likely a good pattern.